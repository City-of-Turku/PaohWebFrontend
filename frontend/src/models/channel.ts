import { Language } from '../types';

export const channelTypes = [
  'EChannel',
  'Phone',
  'PrintableForm',
  'ServiceLocation',
  'WebPage',
] as const;

export type ChannelType = typeof channelTypes[number];

export const isChannelType = (val: string): val is ChannelType => {
  // widen channelTypes to string[] so indexOf(x) works
  return (channelTypes as readonly string[]).indexOf(val) !== -1;
};

export type IChannelDescType = 'Description' | 'Summary';

export interface IChannelDescription {
  value: string;
  type: IChannelDescType;
  language: Language;
}

export interface IPhoneNumber {
  number: string;
  prefixNumber?: string;
  chargeDescription?: string;
  serviceChargeType?: string;
}

export interface IAddress {
  type: string;
  subtype: string;
  streetNumber?: string;
  postalCode?: string;
  latitude?: string;
  longitude?: string;
  streetName?: string;
  postOffice?: string;
  municipalityCode?: string;
  municipalityName?: string;
}

export interface IChannelUrl {
  url: string;
  type: string;
}

export interface IChannel {
  id: string;
  type: ChannelType;
  name: string;
  descriptions: IChannelDescription[];
  webPages: string[];
  emails: string[];
  phoneNumbers: IPhoneNumber[];
  addresses: IAddress[];
  channelUrls: IChannelUrl[];
}
