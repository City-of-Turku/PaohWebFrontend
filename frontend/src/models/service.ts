import { Language } from '../types';

export type ServiceDescriptionType =
  | 'ChargeTypeAdditionalInfo'
  | 'DeadLine'
  | 'Description'
  | 'GD_Description'
  | 'GD_GeneralDescriptionTypeAdditionalInformation'
  | 'GD_Summary'
  | 'ProcessingTime'
  | 'ServiceType'
  | 'Summary'
  | 'UserInstruction'
  | 'ValidityTime';

export interface IDescription {
  value: string;
  type: ServiceDescriptionType;
  language: Language;
  autoTranslated: boolean;
  original?: string; // If name was automatically translated, this contains the original (Finnish)
  originalLanguage?: string;
}

export interface IOrganization {
  id: string | null;
  name: string;
  roleType: OrgRole;
}

export enum OrgRole {
  Responsible = 'Responsible',
  Producer = 'Producer',
  OtherResponsible = 'OtherResponsible',
}

export interface IServiceArea {
  name: string;
  type: string;
  code: string;
}

export interface IServiceName {
  value: string;
  language: Language;
  autoTranslated: boolean;
  original?: string; // If name was automatically translated, this contains the original (Finnish)
  originalLanguage?: Language;
}

export interface IService {
  id: string;
  name: IServiceName;
  descriptions: IDescription[];
  organizations: IOrganization[];
  areas: IServiceArea[];
}
