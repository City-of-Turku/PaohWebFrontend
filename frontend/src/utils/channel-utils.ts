import {
  IChannel,
  IChannelDescription,
  IPhoneNumber,
  IAddress,
} from 'models/channel';

function formatAddressParts(address: IAddress): string[] {
  const streetAddress = address.streetName
    ? `${address.streetName} ${address.streetNumber}`
    : '';
  const municipality = address.municipalityName
    ? `${address.postalCode} ${address.municipalityName}`
    : '';

  return [streetAddress, municipality];
}

function formatAddressOneliner(address: IAddress): string {
  return formatAddressParts(address).join(', ');
}

function formatAddressMultiline(address: IAddress): string {
  return formatAddressParts(address).join('\n');
}

function formatPhoneNumber(number: IPhoneNumber): string {
  return `${number.prefixNumber || ''}${number.number}`;
}

function generateMapsDirectionLink(address: string): string {
  const mapsBaseUrl = 'https://www.google.com/maps/dir/?api=1';
  const destination = '' + encodeURIComponent(address);
  return `${mapsBaseUrl}&destination=${destination}`;
}

function generateMapsLocationLink(longitude: string, latitude: string): string {
  return `https://www.openstreetmap.org/?mlat=${latitude}&mlon=${longitude}&zoom=16`;
}

function getChannelDescription(channel: IChannel): {
  description: string;
  descLang: string | undefined;
} {
  const summary = channel.descriptions.find((d) => d.type == 'Summary');
  const description = channel.descriptions.find(
    (d: IChannelDescription) => d.type == 'Description',
  );

  const chosen = summary || description;

  return {
    description: chosen ? chosen.value : '',
    descLang: chosen ? chosen.language : undefined,
  };
}

function uniqueFormattedAddresses(
  addresses: IAddress[],
  formattingFunc: (a: IAddress) => string,
): string[] {
  const uniqueAddresses: string[] = addresses
    .filter((address: IAddress) => address.type === 'Location')
    .map((a) => formattingFunc(a))
    .filter((formattedAdd) => formattedAdd !== '')
    .filter((addr, index, self) => self.indexOf(addr) === index);

  return uniqueAddresses;
}

export interface IFormmattedLocation {
  address: string;
  latitude: string | undefined;
  longitude: string | undefined;
}

function uniqueFormattedLocationAddresses(
  addresses: IAddress[],
  formattingFunc: (a: IAddress) => string,
): IFormmattedLocation[] {
  const uniqueAddresses: IFormmattedLocation[] = addresses
    .filter((address: IAddress) => address.type === 'Location')
    .map((a) => {
      return {
        address: formattingFunc(a),
        latitude: a.latitude,
        longitude: a.longitude,
      };
    })
    .filter((formattedAdd) => formattedAdd.address !== '')
    .filter(
      (addr, index, self) =>
        self.findIndex((a) => a.address === addr.address) === index,
    );

  return uniqueAddresses;
}

function uniqueFormattedPhoneNumbers(numbers: IPhoneNumber[]): string[] {
  const uniqueNumbers: string[] = numbers
    .map((n) => formatPhoneNumber(n))
    .filter((num, index, self) => self.indexOf(num) === index);
  return uniqueNumbers;
}

export {
  formatAddressMultiline,
  formatAddressOneliner,
  formatPhoneNumber,
  generateMapsDirectionLink,
  generateMapsLocationLink,
  getChannelDescription,
  uniqueFormattedAddresses,
  uniqueFormattedLocationAddresses,
  uniqueFormattedPhoneNumbers,
};
