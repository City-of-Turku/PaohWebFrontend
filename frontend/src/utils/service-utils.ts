import {
  IDescription,
  IOrganization,
  IServiceArea,
  OrgRole,
} from 'models/service';

/* Check if string is valid UUID */
function checkIfValidUUID(str: string): boolean {
  const regexExp =
    /^[0-9a-f]{8}\b-[0-9a-f]{4}\b-[0-9a-f]{4}\b-[0-9a-f]{4}\b-[0-9a-f]{12}$/gi;
  return regexExp.test(str.toLowerCase());
}

/**
 * Order of descriptions to use: Description > GD_Description > Summary > GD_Summary
 * If none of them exist, return an empty description
 */
function getDescription(
  descriptions: IDescription[],
): IDescription | undefined {
  // Convert type: x, value: y structure into {x: y} meaning description types serve as keys for values (descriptions)
  const mappedDescriptions = descriptions.map((desc: IDescription) => ({
    [desc.type]: {
      value: desc.value,
      language: desc.language,
      autoTranslated: desc.autoTranslated,
      ...(desc.original && {
        original: desc.original,
        originalLanguage: desc.originalLanguage,
      }),
    },
  }));
  const descData = Object.assign({}, ...mappedDescriptions);
  const descriptionOrder = [
    'Description',
    'Summary',
    'GD_Description',
    'GD_Summary',
  ];

  const selectedType = descriptionOrder.find(
    (descName: string) =>
      descData[descName] &&
      descData[descName].value !== null &&
      descData[descName].value !== undefined &&
      descData[descName].value !== '',
  );

  if (selectedType) {
    return descData[selectedType];
  }
  return undefined;
}

/**
 *
 * @param areas
 * @param areasText
 * @returns
 */
function getLocationSubheader(
  areas: IServiceArea[],
  areasText: string,
): string | null {
  if (areas && areas.length > 0) {
    return areas.length === 1 ? areas[0].name : `${areas.length} ${areasText}`;
  }
  return null;
}

/**
 * Get name of (one) organization responsible for a service
 * @param organizations Array of organizations
 * @returns name of (first) organization with role Responsible - or if such is not found, returns null
 */
function getOrganizationName(organizations: IOrganization[]): string | null {
  const responsible = organizations.filter(
    (o) => o.roleType === OrgRole.Responsible,
  );
  if (responsible.length > 0) {
    return responsible[0].name;
  }
  return null;
}

export {
  checkIfValidUUID,
  getDescription,
  getLocationSubheader,
  getOrganizationName,
};
