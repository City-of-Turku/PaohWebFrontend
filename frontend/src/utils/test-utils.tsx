import React, { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
import { ThemeProvider as EmotionThemeProvider } from '@emotion/react';
import { ThemeProvider as MuiThemeProvider } from '@material-ui/core';
import { theme } from '../theme/theme';
import { getTranslations } from 'translations/locales';

interface ProviderProps {
  children?: React.ReactNode;
}

function AllTheProviders({ children }: ProviderProps) {
  return (
    <IntlProvider locale="fi" messages={getTranslations('fi')}>
      <MuiThemeProvider theme={theme}>
        <EmotionThemeProvider theme={theme}>{children}</EmotionThemeProvider>
      </MuiThemeProvider>
    </IntlProvider>
  );
}

// eslint-disable-next-line
const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';

export { customRender as render };

/**
 * According to the React docs, Hooks cannot be called from anywhere like regular javascript functions.
 * They can only be called from a functional component or from other hooks.
 *
 * The component below can be used to call hooks in tests. Using a callback, we can pass
 * the result of our custom hook in the outside scope where we want to test it.
 */

/* eslint-disable  @typescript-eslint/no-explicit-any */
const TestHook = ({ callback }: any): null => {
  callback();
  return null;
};

/* eslint-disable  @typescript-eslint/no-explicit-any */
export const testHook = (callback: any): void => {
  render(<TestHook callback={callback} />);
};
