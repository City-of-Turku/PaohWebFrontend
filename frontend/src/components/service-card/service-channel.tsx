/** @jsxRuntime classic */
/** @jsx jsx */
import { css, jsx } from '@emotion/react';
import { FC, Fragment } from 'react';
import { observer } from 'mobx-react-lite';
import { Typography, Box } from '@material-ui/core';
import { IChannel, IChannelUrl, ChannelType } from 'models/channel';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faExternalLinkAlt,
  faFileAlt,
  faGlobe,
  faMapMarkedAlt,
  faMapMarkerAlt,
  faPhone,
} from '@fortawesome/free-solid-svg-icons';
import {
  formatAddressOneliner,
  formatPhoneNumber,
  generateMapsLocationLink,
  uniqueFormattedLocationAddresses,
  IFormmattedLocation,
} from '../../utils/channel-utils';
import { StyledLink } from '../link/link';
import { useIntl } from 'react-intl';

interface IServiceChannel {
  channel: IChannel;
  type: ChannelType;
}

export const ServiceChannel: FC<IServiceChannel> = observer(
  ({ channel, type }) => {
    if (type === 'EChannel' || type === 'WebPage') {
      if (channel.webPages.length === 0) {
        return null;
      } else {
        return (
          <li>
            <StyledLink
              url={channel.webPages[0]}
              external
              content={channel.name}
              iconBefore={faGlobe}
              iconAfter={faExternalLinkAlt}
            />
          </li>
        );
      }
    }

    if (type === 'Phone') {
      if (channel.phoneNumbers.length > 0) {
        return (
          <li>
            <StyledLink
              url={`tel:${formatPhoneNumber(channel.phoneNumbers[0])}`}
              external
              content={formatPhoneNumber(channel.phoneNumbers[0])}
              iconBefore={faPhone}
            />{' '}
            {channel.name}
          </li>
        );
      } else {
        return null;
      }
    }

    if (type === 'PrintableForm') {
      if (channel.channelUrls.length > 0) {
        return (
          <Fragment>
            {channel.channelUrls.map((form: IChannelUrl) => (
              <li key={form.url}>
                <StyledLink
                  external
                  url={form.url}
                  content={`${channel.name} (${form.type})`}
                  iconBefore={faFileAlt}
                />
              </li>
            ))}
          </Fragment>
        );
      }
      return null;
    }

    if (type === 'ServiceLocation') {
      return (
        <li
          className="service-location"
          css={css`
            .service-location-name {
              font-weight: 600;
            }
          `}
        >
          <ServiceLocationName channel={channel} />
          <LocationAddress channel={channel} />
        </li>
      );
    }

    return null;
  },
);

const LocationAddress: FC<{ channel: IChannel }> = observer(({ channel }) => {
  const intl = useIntl();

  if (channel.addresses.length == 0) return null;
  const addresses = uniqueFormattedLocationAddresses(
    channel.addresses,
    formatAddressOneliner,
  );

  if (addresses.length == 0) return null;

  return (
    <Fragment>
      {addresses.map((add: IFormmattedLocation) => (
        <Box key={add.address} className="channel-address">
          <Typography variant="body1">{add.address}</Typography>
          {add.latitude && add.longitude && (
            <StyledLink
              url={generateMapsLocationLink(add.longitude, add.latitude)}
              external
              content={intl.formatMessage({
                id: 'ShowLocationOnMap',
                defaultMessage: 'Sijainti kartalla',
                description:
                  "Text on a link that opens map service showing service's location",
              })}
              iconBefore={faMapMarkedAlt}
            />
          )}
        </Box>
      ))}
    </Fragment>
  );
});

/**
 * Format name of a service as a link if such information is available.
 * Otherwise return just a typography element
 * @param ch  Service channel
 * @returns
 */
const ServiceLocationName: FC<{ channel: IChannel }> = observer(
  ({ channel }) => {
    if (channel.webPages.length > 0) {
      return (
        <StyledLink
          className="service-location-name"
          external
          url={channel.webPages[0]}
          content={channel.name}
          iconBefore={faMapMarkerAlt}
          css={{ fontWeight: 700 }}
        />
      );
    }

    return (
      <div
        css={css`
          display: inline-flex;
          align-items: center;

          .fontawesome-icon {
            margin-right: 0.5rem;
            height: 1rem;
            width: 1rem;
          }
        `}
      >
        <FontAwesomeIcon icon={faMapMarkerAlt} className="fontawesome-icon" />
        <Typography className="service-location-name" variant="body1">
          {channel.name}
        </Typography>
      </div>
    );
  },
);
