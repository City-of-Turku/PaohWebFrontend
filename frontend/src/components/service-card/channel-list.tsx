/** @jsxRuntime classic */
/** @jsx jsx */
import { css, jsx } from '@emotion/react';
import { FC } from 'react';
import { observer } from 'mobx-react-lite';
import { Typography, Box } from '@material-ui/core';
import { ChannelType, channelTypes, IChannel } from 'models/channel';
import { useIntl } from 'react-intl';
import { ServiceChannel } from './service-channel';
import { StyledLink } from '../link/link';
import { faArrowRight } from '@fortawesome/free-solid-svg-icons';

const MAX_CHANNELS_TO_SHOW = 3;
interface IServiceChannelList {
  serviceId: string;
  channels: IChannel[];
}

interface IServiceChannelGroup {
  serviceId: string;
  group: ChannelType;
  channels: IChannel[];
}

export const ServiceChannelList: FC<IServiceChannelList> = observer(
  ({ serviceId, channels }) => {
    return (
      <Box
        css={css`
          > div {
            padding-bottom: 0.5rem;
          }

          > div:last-of-type {
            padding-bottom: 0;
          }
        `}
      >
        {channelTypes.map((ct: ChannelType) => (
          <ServiceChannelGroup
            key={ct}
            group={ct}
            channels={channels}
            serviceId={serviceId}
          />
        ))}
      </Box>
    );
  },
);

export const ServiceChannelGroup: FC<IServiceChannelGroup> = observer(
  ({ group, channels, serviceId }) => {
    const intl = useIntl();
    const channelsByType: IChannel[] = channels.filter((c) => c.type === group);

    // Don't render a section if there are no channels of this type
    if (!channels || channelsByType.length === 0) {
      return null;
    }

    return (
      <Box
        css={css`
          && {
            &:last-of-type {
              padding-bottom: 1rem;
            }

            ul {
              margin-top: 0rem;
              list-style: none;
              padding-inline-start: 0;
              margin-block-end: 0;
            }

            h6 {
              margin-top: 0.75rem;
              text-transform: uppercase;
              margin-bottom: 0.25rem;
            }

            .channel-address {
              margin-left: 1.5rem;
            }

            li.service-location {
              padding-bottom: 1.2rem;

              p {
                line-height: inherit;
              }
            }

            li.service-location:last-of-type {
              padding-bottom: 0.5rem;
            }
          }
        `}
      >
        <Typography variant="h6">
          {intl.formatMessage({
            id: `ChannelGroupTitle${group}`,
            defaultMessage: group,
          })}
          {` (${channelsByType.length} ${intl.formatMessage({
            id: 'Pieces',
            defaultMessage: group,
          })})`}
        </Typography>
        <ul>
          {channelsByType.slice(0, MAX_CHANNELS_TO_SHOW).map((c) => (
            <ServiceChannel key={c.id} channel={c} type={group} />
          ))}
          {channelsByType.length > MAX_CHANNELS_TO_SHOW && (
            <StyledLink
              url={`${serviceId}?ct=${group}`}
              content={intl.formatMessage({
                id: `ShowAll${group}`,
                defaultMessage: 'Näytä kaikki',
              })}
              iconBefore={faArrowRight}
              css={css`
                font-weight: 700;
                margin: 0.5rem;
                margin-left: 1.5rem;
              `}
            />
          )}
        </ul>
      </Box>
    );
  },
);
