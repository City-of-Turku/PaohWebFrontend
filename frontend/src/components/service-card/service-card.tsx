/** @jsxRuntime classic */
/** @jsx jsx */
import { css, jsx } from '@emotion/react';
import { FC, useState } from 'react';
import { observer } from 'mobx-react-lite';
import {
  Card,
  CardActions,
  CardHeader,
  CardContent,
  Collapse,
  IconButton,
  Typography,
  useTheme,
} from '@material-ui/core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faAngleDown,
  faAngleUp,
  faLanguage,
} from '@fortawesome/free-solid-svg-icons';
import { IDescription, IService } from 'models/service';
import { IChannel } from 'models/channel';
import { ServiceChannelList } from './channel-list';
import { useIntl } from 'react-intl';
import { getDescription } from '../../utils/service-utils';
import { ServiceName, ServiceDescriptionBox } from '../service-atoms';
import { OnOffButton } from 'components/buttons/on-off-button';

const DESCRIPTION_PREVIEW_LIMIT = 300;

interface IServiceCard {
  service: IService;
  channels: IChannel[];
}

export const ServiceCard: FC<IServiceCard> = observer(
  ({ service, channels }) => {
    const theme = useTheme();
    const intl = useIntl();
    const [expanded, setExpanded] = useState(false);
    const [autoTranslate, setAutoTranslate] = useState(true);

    const handleExpandClick = () => {
      setExpanded(!expanded);
    };

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const toggleAutoTranslate = (id: string) => {
      setAutoTranslate((prevState) => !prevState);
    };

    const descObject: IDescription | undefined = getDescription(
      service.descriptions,
    );

    const hasAutoTranslatedContent =
      descObject?.autoTranslated || service.name.autoTranslated;

    const description = autoTranslate
      ? descObject?.value || ''
      : descObject?.original || '';

    // Show action buttons only if there is more content to show
    const moreDetailsToShow: boolean =
      description.length > DESCRIPTION_PREVIEW_LIMIT || channels.length > 0;

    return (
      <Card
        css={css`
          && {
            border-radius: 1rem;
            background-color: ${theme.palette.section.main};
            box-shadow: none;
            padding: ${theme.spacing(2)}px;

            .MuiCardContent-root,
            .MuiCardActions-root,
            .MuiCardHeader-root {
              padding: 0;
            }

            .MuiCardContent-root {
              padding-top: ${theme.spacing(0)}px;
              padding-bottom: ${theme.spacing(0)}px;
            }

            .MuiCardHeader-root {
              padding-bottom: ${theme.spacing(1)}px;
            }

            .MuiCardHeader-title {
              & a {
                padding: 0.25rem 0;
              }

              .service-name {
                font-size: 1.35rem;
                line-height: 1.35rem;
              }
            }

            .MuiCardActions-root {
              padding-top: ${theme.spacing(1)}px;
            }

            .MuiCardActions-root .MuiIconButton-root {
              border-radius: 1rem;
              padding: 0;
              background: inherit;
              display: flex;
              flex-direction: row;
              align-items: center;
              line-height: 1.75rem;
              font-size: 1rem;

              .MuiTypography-root {
                font-weight: 700;
              }

              .fontawesome-icon {
                color: ${theme.palette.text.primary};
                height: 1.4rem;
                width: 1.4rem;
                margin-left: 0.15rem;
              }
            }

            .MuiButton-root {
              color: ${theme.palette.primary.main};
            }

            .description h3 {
              font-size: 1.025rem;
              line-height: 1.25rem;
              font-weight: 500;
              margin-block-start: 0.75rem; // Space above title
              margin-block-end: 0rem; // Space below title

              &:first-of-type {
                margin-block-start: 0rem;
                margin-block-end: 0.5rem;
              }
            }

            .description p {
              padding-top: 0.75rem;
              margin-block-start: 0;
              margin-block-end: 0;
            }

            // Tighter spacing if paragrahs comes after title
            .description h3 + p {
              padding-top: 0.5rem;
            }

            .description p:first-of-type {
              padding-top: 0;
            }

            .description ul,
            .description ol {
              margin-block-start: 0;
              margin-block-end: 0;
              margin-top: 0;
              padding-inline-start: 2rem;
            }
          }
        `}
      >
        <CardHeader
          title={
            <div
              css={css`
                display: flex;
                align-items: flex-start;
                justify-content: space-between;
              `}
            >
              <ServiceName
                service={service}
                autoTranslate={autoTranslate}
                nameIsLink={true}
              />
              {hasAutoTranslatedContent && (
                <OnOffButton
                  id="toggle-autotranslate"
                  on={autoTranslate}
                  onClick={toggleAutoTranslate}
                  icon={faLanguage}
                  showTooltip
                  offTooltipTextId={'AutomaticTranslationIsOff'}
                  onTooltipTextId={'AutomaticTranslationIsOn'}
                />
              )}
            </div>
          }
        />
        <CardContent>
          <ServiceDescriptionBox
            description={descObject}
            autoTranslate={autoTranslate}
            expanded={expanded}
            collapseLength={DESCRIPTION_PREVIEW_LIMIT}
          />
        </CardContent>
        <CardContent>
          <Collapse in={expanded} timeout="auto" unmountOnExit>
            <ServiceChannelList serviceId={service.id} channels={channels} />
          </Collapse>
        </CardContent>
        {moreDetailsToShow && (
          <CardActions>
            <IconButton
              onClick={handleExpandClick}
              aria-expanded={expanded}
              aria-label={expanded ? 'show less' : 'show more'}
            >
              <Typography variant="body2" color="textPrimary" component="span">
                {expanded
                  ? intl.formatMessage({
                      id: 'CardShowLess',
                      defaultMessage: 'Lue v채hemm채n',
                    })
                  : intl.formatMessage({
                      id: 'CardShowMore',
                      defaultMessage: 'Lue lis채채',
                    })}
              </Typography>
              {expanded ? (
                <FontAwesomeIcon
                  icon={faAngleUp}
                  className="fontawesome-icon"
                />
              ) : (
                <FontAwesomeIcon
                  icon={faAngleDown}
                  className="fontawesome-icon"
                />
              )}
            </IconButton>
          </CardActions>
        )}
      </Card>
    );
  },
);
