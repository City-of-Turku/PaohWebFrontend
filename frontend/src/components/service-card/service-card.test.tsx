import React from 'react';
import { render, screen } from '../../utils/test-utils';
import { IRecommendation } from '../../models/recommendation';
import { IService } from '../../models/service';
import { ServiceCard } from './service-card';
import { MemoryRouter } from 'react-router-dom';

const testRec: IRecommendation = {
  service: {
    id: '1',
    name: {
      value: 'Esimerkkipalvelu',
      language: 'fi',
      autoTranslated: false,
    },
    organizations: [],
    descriptions: [
      {
        value: 'Tämä on palvelun kuvaus.',
        type: 'Description',
        language: 'fi',
        autoTranslated: false,
      },
      {
        value: 'Tämä on tiivistelmä palvelun kuvauksesta.',
        type: 'Summary',
        language: 'fi',
        autoTranslated: false,
      },
    ],
    areas: [],
  },
  channels: [],
};

test('renders a recommendation that does not have any channels', () => {
  const { asFragment } = render(
    <MemoryRouter>
      <ServiceCard service={testRec.service} channels={testRec.channels} />
    </MemoryRouter>,
  );
  expect(asFragment()).toMatchSnapshot();
});

describe('Extracting description from a service', () => {
  const serviceWithoutDescription: IService = {
    id: '404',
    name: {
      value: 'Esimerkkipalvelu',
      language: 'fi',
      autoTranslated: false,
    },
    organizations: [],
    descriptions: [],
    areas: [],
  };

  test('renders a service that does not have any description available', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <ServiceCard service={serviceWithoutDescription} channels={[]} />
      </MemoryRouter>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  const serviceWithSummary: IService = {
    id: '404',
    name: {
      value: 'Esimerkkipalvelu',
      language: 'fi',
      autoTranslated: false,
    },
    organizations: [],
    descriptions: [
      {
        value: 'Tämä on palvelun GD_Description.',
        type: 'GD_Description',
        language: 'fi',
        autoTranslated: false,
      },
      {
        value: 'Tämä on tiivistelmä palvelun kuvauksesta.',
        type: 'Summary',
        language: 'fi',
        autoTranslated: false,
      },
    ],
    areas: [],
  };

  const allDescAvailable: IService = {
    id: '100',

    name: {
      value: 'Esimerkkipalvelu kaikilla kuvauksilla',
      language: 'fi',
      autoTranslated: false,
    },
    organizations: [],
    descriptions: [
      {
        value: 'Tämä on palvelun kuvaus.',
        type: 'Description',
        language: 'fi',
        autoTranslated: false,
      },
      {
        value: 'Tämä on tiivistelmä palvelun kuvauksesta.',
        type: 'Summary',
        language: 'fi',
        autoTranslated: false,
      },
      {
        value: 'Tämä on palvelun GD_Description.',
        type: 'GD_Description',
        language: 'fi',
        autoTranslated: false,
      },
      {
        value: 'Tämä on palvelun GD_Summary.',
        type: 'GD_Summary',
        language: 'fi',
        autoTranslated: false,
      },
    ],
    areas: [],
  };

  it('Show service description despite other descriptions being available', () => {
    render(
      <MemoryRouter>
        <ServiceCard service={allDescAvailable} channels={[]} />
      </MemoryRouter>,
    );

    const description = screen.getByText('Tämä on palvelun kuvaus.');
    expect(description).toBeInTheDocument();

    expect(
      screen.queryByText('Tämä on tiivistelmä palvelun kuvauksesta.'),
    ).toBeNull();
  });

  it('Shows a placeholder/error message when description missing', () => {
    render(
      <MemoryRouter>
        <ServiceCard service={serviceWithoutDescription} channels={[]} />
      </MemoryRouter>,
    );
    const errorMsg = screen.getByText(
      'Virhe: Palvelulle ei ole saatavilla kuvausta.',
    );
    expect(errorMsg).toBeInTheDocument();
  });

  it('Shows the summary if description not available', () => {
    render(
      <MemoryRouter>
        <ServiceCard service={serviceWithSummary} channels={[]} />
      </MemoryRouter>,
    );
    const summaryText = screen.getByText(
      'Tämä on tiivistelmä palvelun kuvauksesta.',
    );
    expect(summaryText).toBeInTheDocument();
  });

  const serviceWithGD: IService = {
    id: '404',
    name: {
      value: 'Esimerkkipalvelu',
      language: 'fi',
      autoTranslated: false,
    },
    organizations: [],
    descriptions: [
      {
        value: 'Tämä on palvelun GD_Description.',
        type: 'GD_Description',
        language: 'fi',
        autoTranslated: false,
      },
      {
        value: 'Tämä on palvelun GD_Summary.',
        type: 'GD_Summary',
        language: 'fi',
        autoTranslated: false,
      },
    ],
    areas: [],
  };

  it('Shows general description if summary or description is not available', () => {
    render(
      <MemoryRouter>
        <ServiceCard service={serviceWithGD} channels={[]} />
      </MemoryRouter>,
    );
    expect(screen.getByText(/GD_Description/)).toBeInTheDocument();
  });

  const serviceWithGDSummary: IService = {
    id: '404',
    name: {
      value: 'Esimerkkipalvelu',
      language: 'fi',
      autoTranslated: false,
    },
    organizations: [],
    descriptions: [
      {
        value: 'Tämä on palvelun GD_Summary.',
        type: 'GD_Summary',
        language: 'fi',
        autoTranslated: false,
      },
    ],
    areas: [],
  };

  it('Shows general summary if nothing else is available', () => {
    render(
      <MemoryRouter>
        <ServiceCard service={serviceWithGDSummary} channels={[]} />
      </MemoryRouter>,
    );
    expect(screen.getByText(/GD_Summary/)).toBeInTheDocument();
  });
});
