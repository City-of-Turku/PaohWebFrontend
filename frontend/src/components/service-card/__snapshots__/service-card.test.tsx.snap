// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Extracting description from a service renders a service that does not have any description available 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root css-1iszrn9 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
        >
          <div
            class="css-oj1ggt"
          >
            <div
              class="css-14yu5aj"
            >
              <h3
                class="MuiTypography-root organization-name MuiTypography-body1"
              />
              <h2
                class="MuiTypography-root service-name MuiTypography-h1"
                lang="fi"
              >
                <a
                  class="css-shy10a"
                  href="/404"
                >
                  <span
                    class="link-content"
                  >
                    Esimerkkipalvelu
                  </span>
                </a>
              </h2>
              <h3
                class="MuiTypography-root service-area MuiTypography-body1"
              />
            </div>
          </div>
        </span>
      </div>
    </div>
    <div
      class="MuiCardContent-root"
    >
      <div>
        Virhe: Palvelulle ei ole saatavilla kuvausta.
      </div>
    </div>
    <div
      class="MuiCardContent-root"
    />
  </div>
</DocumentFragment>
`;

exports[`renders a recommendation that does not have any channels 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root css-1iszrn9 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
        >
          <div
            class="css-oj1ggt"
          >
            <div
              class="css-14yu5aj"
            >
              <h3
                class="MuiTypography-root organization-name MuiTypography-body1"
              />
              <h2
                class="MuiTypography-root service-name MuiTypography-h1"
                lang="fi"
              >
                <a
                  class="css-shy10a"
                  href="/1"
                >
                  <span
                    class="link-content"
                  >
                    Esimerkkipalvelu
                  </span>
                </a>
              </h2>
              <h3
                class="MuiTypography-root service-area MuiTypography-body1"
              />
            </div>
          </div>
        </span>
      </div>
    </div>
    <div
      class="MuiCardContent-root"
    >
      <div
        class="css-1anmknu"
      >
        <div
          class="description"
          lang="fi"
        >
          Tämä on palvelun kuvaus.
        </div>
      </div>
    </div>
    <div
      class="MuiCardContent-root"
    />
  </div>
</DocumentFragment>
`;
